x/accounts/vesting/vesting.go

```diff
@@ -0,0 +1,490 @@
+	DELAYED_VESTING_ACCOUNT    = "delayed-vesting-account"
+	PERIODIC_VESTING_ACCOUNT   = "periodic-vesting-account"
+	PERMANENT_VESTING_ACCOUNT  = "permanent-vesting-account"
+)
+
+type getVestingFunc = func(ctx context.Context, time time.Time) (sdk.Coins, error)
+
+// NewBaseVesting creates a new BaseVesting object.
+func NewBaseVesting(d accountstd.Dependencies) *BaseVesting {
+	baseVesting := &BaseVesting{
+		Owner:            collections.NewItem(d.SchemaBuilder, OwnerPrefix, "owner", collections.StringValue),
+		OriginalVesting:  collections.NewMap(d.SchemaBuilder, OriginalVestingPrefix, "original_vesting", collections.StringKey, sdk.IntValue),
+		DelegatedFree:    collections.NewMap(d.SchemaBuilder, DelegatedFreePrefix, "delegated_free", collections.StringKey, sdk.IntValue),
+		DelegatedVesting: collections.NewMap(d.SchemaBuilder, DelegatedVestingPrefix, "delegated_vesting", collections.StringKey, sdk.IntValue),
+		addressCodec:     d.AddressCodec,
+		headerService:    d.HeaderService,
+		EndTime:          collections.NewItem(d.SchemaBuilder, EndTimePrefix, "end_time", sdk.IntValue),
+	}
+
+	return baseVesting
+}
+
+type BaseVesting struct {
+	// Owner is the address of the account owner.
+	Owner            collections.Item[string]
+	OriginalVesting  collections.Map[string, math.Int]
+	DelegatedFree    collections.Map[string, math.Int]
+	DelegatedVesting collections.Map[string, math.Int]
+	addressCodec     address.Codec
+	headerService    header.Service
+	// Vesting end time, as unix timestamp (in seconds).
+	EndTime collections.Item[math.Int]
+}
+
+func (bva *BaseVesting) Init(ctx context.Context, msg *vestingtypes.MsgInitVestingAccount) (
+	*vestingtypes.MsgInitVestingAccountResponse, error,
+) {
+	sender := accountstd.Sender(ctx)
+	if sender == nil {
+		return nil, sdkerrors.ErrInvalidAddress.Wrap("Cannot find sender address from context")
+	}
+	to := accountstd.Whoami(ctx)
+	if to == nil {
+		return nil, sdkerrors.ErrInvalidAddress.Wrap("Cannot find account address from context")
+	}
+
+	toAddress, err := bva.addressCodec.BytesToString(to)
+	if err != nil {
+		return nil, sdkerrors.ErrInvalidAddress.Wrapf("invalid 'to' address: %s", err)
+	}
+	fromAddress, err := bva.addressCodec.BytesToString(sender)
+	if err != nil {
+		return nil, sdkerrors.ErrInvalidAddress.Wrapf("invalid 'sender' address: %s", err)
+	}
+	err = bva.Owner.Set(ctx, msg.Owner)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := validateAmount(msg.Amount); err != nil {
+		return nil, err
+	}
+
+	sortedAmt := msg.Amount.Sort()
+	for _, coin := range sortedAmt {
+		err = bva.OriginalVesting.Set(ctx, coin.Denom, coin.Amount)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	err = bva.EndTime.Set(ctx, math.NewInt(msg.EndTime))
+	if err != nil {
+		return nil, err
+	}
+
+	// Send token to new vesting account
+	sendMsg := banktypes.NewMsgSend(fromAddress, toAddress, msg.Amount)
+	if _, err = accountstd.ExecModule[banktypes.MsgSendResponse](ctx, sendMsg); err != nil {
```

Comment:
this will not work as an x/accounts is not allowed to impersonate anybody, in this case the x/account "toAddress" is trying to impersonate "fromAddress"

Reply:
The fromAddress should fund the account separetedly (with another message after init)

Reply:
ahhhh the sender is always the x/accounts account

Reply:
remove this for the future x/accounts funding feature